Api for notifications

fabrixxm opened the pull request:
and a new class "NotificationsManager"

annando commented:
Could you extend the API documentation in doc/api.md?

rabuzarus commented on include/NotificationsManager.php:
```diff
@@ -0,0 +1,113 @@
+<?php
+require_once("include/datetime.php");
```
```
/**
 * @file include/NotificationsManager.php
 */
```

rabuzarus replied:
```
/**
 * @file include/NotificationsManager.php
 */
```
above require_once

rabuzarus commented on include/NotificationsManager.php:
```diff
@@ -0,0 +1,113 @@
+require_once("include/datetime.php");
+
+class NotificationsManager {
+    private $a;
```
```
/**
 * @brief Some description of the class
 */

rabuzarus commented on include/NotificationsManager.php:
```diff
@@ -0,0 +1,113 @@
+<?php
+require_once("include/datetime.php");
+
```
```
/**
 * @brief Some description of the class
 */
```

rabuzarus commented on include/NotificationsManager.php:
```diff
@@ -0,0 +1,113 @@
+     * @return bool true on success, false on errors
+     */
+    public function setSeen($note, $seen = true) {
+        return q("update notify set seen = %d where ( link = '%s' or ( parent != 0 and parent = %d and otype = '%s' )) and uid = %d",
```
Should follow the unofficial coding standard for SQL queries 
```
"UPDATE `notify` SET `seen` = %d WHERE ( `link` = '%s' OR ( `parent` != 0 AND `parent` = %d AND `otype` = '%s' )) AND `uid` = %d",
```
maybe some line breaks would be better for readability

rabuzarus commented on include/NotificationsManager.php:
```diff
@@ -0,0 +1,113 @@
+     * @return array note values or null if not found
+     */
+    public function getByID($id) {
+        $r = q("select * from notify where id = %d and uid = %d limit 1",
```
Should follow the unofficial friendica coding standard for SQL queries
```
"SELECT * FROM `notify` WHERE `id` = %d AND `uid` = %d LIMIT 1",
```

rabuzarus commented on include/NotificationsManager.php:
```diff
@@ -0,0 +1,113 @@
+     * @return bool true on success, false on error
+     */
+    public function setAllSeen($seen = true) {
+    	return q("update notify set seen = %d where uid = %d",
```
Should follow the unofficial friendica coding standard for SQL queries:
```
"UPDATE `notify` SET `seen` = %d WHERE `uid` = %d"
 ```

rabuzarus commented on include/api.php:
```diff
@@ -3386,6 +3414,43 @@ function api_friendica_activity(&$a, $type) {
+	/**
+	 * returns notifications
+	 * if called with note id set note seen and returns associated item (if possible)
+	 */
```
```
/**
 * @brief Returns notifications
 *
 * If called with note id set note seen and returns associated item (if possible)
 *
 * @param App $a
 * @param string $type Known types are 'atom', 'rss', 'xml' and 'json'
 */
```
I don't know if we could give a good `@return` description here

rabuzarus commented on include/NotificationsManager.php:
```diff
@@ -0,0 +1,113 @@
+        $this->a = get_app();
+    }
+    
+    private function _set_extra($notes) {
```
This should have a short doxygen documentation if possible

fabrixxm commented:
I've added doxygen comments as @rabuzarus pointed and I've updated docs.
I've separated code for list notifications and for set seen, to set different required http method.
I've not added backticks to queries because is a mysql-only thing to escape problematic names, and I think that if it's not needed it shouldn't be used... but if you think is worth add them, I'll add them.

rabuzarus commented:
@fabrixxm 
According to the sql queries. It would be really good to have a consistent style here. There are two reasons from my side:
* If you working on a project where you often read much code you get used to this style. Another style make it harder to read. But the more important point is,
* if you edit something somewhere else in the code or at the db scheme (or sometimes only because you want to understand something) you search for code snippet (like with grep). It is very hard to find the needed piece of code by searching if we have too many possibilities how it could look like. We should avoid this where we can. This could lower developing time and some bugs

fabrixxm commented:
:+1:

rabuzarus commented on include/api.php:
```diff
@@ -3386,6 +3419,64 @@ function api_friendica_activity(&$a, $type) {
+		$nm->setSeen($note);
+		if ($note['otype']=='item') {
+			// would be really better with an ItemsManager and $im->getByID() :-P
+			$r = q("SELECT * FROM item WHERE id=%d AND uid=%d",
```
```
"SELECT * FROM `item` WHERE `id` = %d AND `uid` = %d",
```

annando replied:
I, personally, would prefer selecting only the needed fields. This makes queries faster and wastes fewer memory.

fabrixxm replied:
That line returns an API item object. I could replace that with a query similar to one in `api_statuses_user_timeline()`, on line 1769, but it has "SELECT `item`.* [...]" that too...
Could be a point for a future PR, maybe when introducing an ItemsManager class...

rabuzarus commented on mod/notify.php:
```diff
@@ -47,38 +38,37 @@ function notify_init(&$a) {
+	if ($r!==false && count($r) > 0) {
+		foreach ($r as $it) {
+			$notif_content .= replace_macros($not_tpl,array(
+				'$item_link' => $a->get_baseurl(true).'/notify/view/'. $it['id'],
```
just a hint:
we can now use `app::get_baseurl()` instead of `$a->get_baseurl()`  so we don't have to transmit `$a` when the only thing we need from `$a` would be the baseurl

fabrixxm replied:
that line hasn't been modified, just the indentation. I would leave this kind of fix for another PR.

But anyway, that code is in `mod/notify.php::notify_content(&$a)`, which already get the `$a` value...

annando commented:
Could you resolve the conflict? :-)

fabrixxm commented:
oups ! :-)

fabrixxm commented:
ready

tobiasd commented on include/NotificationsManager.php:
```diff
@@ -0,0 +1,135 @@
+            $n['timestamp'] = strtotime($local_time);
+            $n['date_rel'] = relative_date($n['date']);
+			$n['msg_html'] = bbcode($n['msg'], false, false, false, false);
+			$n['msg_plain'] = explode("\n",trim(html2plain($n['msg_html'], 0)))[0];
```
the html2plain function is unknown at this point on my node which results in a fatal error when following a notification to a posting.

fabrixxm replied:
should be fixed by PR #2343

tobiasd replied:
it is :-)
